File: Semestre_2/Vision_Computadora/Entregas/Matlab/T2_8_Umbralizazion_Segmentacion_Fondo.m
close all;
clc;

[file,path] = uigetfile({'*.jpg';'*.png';'*.jpeg';'*.bmp';'*.gif'}, 'File Selector');
ruta = fullfile(path, file);

I = imread(ruta);
I = im2double(I);

% Pasar a gris sólo si es RGB
if length(size(I)) == 3
    A = rgb2gray(I);
else
    A = I;
end

[ax, ay] = size(A);

% Umbrales
u1 = 42/255;
u2 = 180/255;

R1 = zeros(size(A));
R2 = zeros(size(A));

for m = 1:ax
    for n = 1:ay
        if A(m,n) > u1
            R1(m,n) = 1;
        else
            R1(m,n) = 0;
        end

        if A(m,n) > u2
            R2(m,n) = 1;
        else
            R2(m,n) = 0;
        end
    end
end

% Cortar rango (192,255)
low = 192/255;
high = 255/255;

mask = (A >= low) & (A <= high);   % máscara en gris

C = I;
if length(size(I)) == 3
    % aplicar máscara a los 3 canales
    C(repmat(~mask,[1 1 3])) = 0;   % fondo negro
else
    C(~mask) = 0;
end

figure;

subplot(2,2,1);
imshow(I);
title('Imagen de entrada');

subplot(2,2,2);
imshow(R1);
title('Umbralizar, u = 42');

subplot(2,2,3);
imshow(R2);
title('Umbralizar, u = 180');

subplot(2,2,4);
imshow(C);
title('Cortar rango (192, 255)');

File: Semestre_2/Vision_Computadora/Entregas/Python/T2_8_Umbralizazion_Segmentacion_Fondo.py
import matplotlib.pyplot as plt
import numpy as np
from skimage.io import imread
from skimage.color import rgb2gray
from tkinter import Tk, filedialog

Tk().withdraw()
file_path = filedialog.askopenfilename(
    initialdir="C:\\Users\\death\\Documents\\Maestria\\MCIA\\Semestre_2\\Vision_Computadora\\img",
    filetypes=[("Images", "*.jpg;*.png;*.jpeg;*.bmp;*.gif")]
)

I = imread(file_path)

if I.ndim == 3 and I.shape[-1] == 4:
    I = I[..., :3]

I = I.astype(float) / 255.0

if I.ndim == 3:
    A = rgb2gray(I)
else:
    A = I

ax, ay = A.shape

u1 = 42 / 255.0
u2 = 180 / 255.0

R1 = np.zeros_like(A)
R2 = np.zeros_like(A)

R1[A > u1] = 1.0
R2[A > u2] = 1.0

low = 192 / 255.0
high = 255 / 255.0

mask = (A >= low) & (A <= high)

C = I.copy()
if I.ndim == 3:
    C[~mask] = 0.0
else:
    C[~mask] = 0.0

plt.figure()

plt.subplot(2,2,1); plt.imshow(I, cmap=None if I.ndim == 3 else 'gray'); plt.title('Imagen de entrada'); plt.axis('off')
plt.subplot(2,2,2); plt.imshow(R1, cmap='gray'); plt.title('Umbralizar, u = 42'); plt.axis('off')
plt.subplot(2,2,3); plt.imshow(R2, cmap='gray'); plt.title('Umbralizar, u = 180'); plt.axis('off')
plt.subplot(2,2,4); plt.imshow(C, cmap=None if I.ndim == 3 else 'gray'); plt.title('Cortar rango (192, 255)'); plt.axis('off')

plt.show()